GenericModules.md

```diff
@@ -0,0 +1,217 @@
+# Generics Modules for WGSL
+
+# Summary
+
+We propose adding a mechanism to allow generic programming using WESL modules as an extension to the language.
+
+## Assumptions
+
+Assumes that [`load`](./Imports.md), [Modules](./Modules.md) and [Module Interfaces](./ModulesInterfaces.md) have already been implemented. 
+
+# Motivation
+
+Generic programming is useful for WESL, particularly for libraries.
+
+With generic modules, operations like `Reduce` or `PrefixSum` wouldn't need to be manually rewritten for each combination of element type and binary operation.
+
+Larger projects like GPU accelerated particle systems could also benefit from generic modules. This is because the 
+encapsulation of types, functions and other globals in modules provides a means of structuring libraries in a way that is 
+readily user extensible, reusable, and maintainable. 
+
+
+# Guide-level explanation
+
+Generic modules and signatures are declared using the `mod` keyword and like generic types and built-in functions, use angle brackets (`<>`) to denote the generic parameters. Parameters are permitted to be other modules. 
+
+An insantiation of a generic module can be declared inline when used, added to the current namespace using [`include`](./Include.md) (if implemented) or aliased to give the module a concrete name. 
+
+In addition to generic modules, this proposal also requires that [module signatures](./ModulesInterfaces.md) support generic parameters. Module signatures may be used to constrain the type of a generic argument.
+Generic module signatures in generic type constraints may additionally use `_` as a "hole" in arguments to indicate that the user doesn't care about the type of a particular generic parameter. 
+
+Below is an annotated example of how generic modules may be used in practise. This has been translated from the 
+[StoneBerry WebGPU Repository](https://github.com/stoneberry-webgpu/) into the proposed WESL format
+
+```rescript
+// Module signature that simply exposes the single type T. Could perhaps later be sugared to elide the module in follow-up 
+// work
+mod sig Type {
+  type T;
+}
+
+// Module signature that exposes a single constant `value`. Could perhaps later be sugared to elide the module in follow-up work
+mod sig Const<Type: Type> {
+  const value: Type::T;
+}
+
+// Abstract representation of a binary operation. 
+mod sig BinaryOp<OpElem: Type, LoadElem: Type> {
```

Comment:
Do you think generics w/o type bounds are a useful intermediate step? 
I'm not certain, but I'd been hoping so to make things easier. 
(w/o type constraints generics would be more about getting the flexibility to replace the string templating that current wgsl users do. But that weaker form wouldn't add type safety, so some errors would be caught later perhaps by the dawn/wgpu parser)

Reply:
It might be but would entail a breaking change once support for constraints were added.

Reply:
why necessarily a breaking change? I'd think simple substitutions e.g. f32 for LoadElem would be sound even when better typechecking comes in later. Inference would reduce boilerplate, but not render the boilerplate unsound.. hmm.. I bet there's a case I'm missing,  can you explain?

Reply:
@mighdoll The problem is that this design does not allow you to access any members of the module were the generic constraint omitted (which is why it's not permitted to omit the constraint in the grammar in the first place). So if you added typechecking after the fact, you'd get a lot of errors.

Reply:
This is because this design for simplicity does not use type inference

Reply:
Let me restate to see if I understand:
. if we were to release an edition with a generics-light (no contraints, no typechecking, no inference)
. it's perhaps possible to add typechecking in a later edition in principle, but
. if the next stage is generics-checked (constraints, typechecking, no inference)
. generics-checked will want to require explicit constraints for every type parameter (because it doesn't have inference)  
. so best not to allow unconstrained type parameters, lest generics-light code become incompatible with generics-checked

Reply:
As mentioned on discord, there is one way to preserve compatibility in this design which is to add the concept of an "any" module type, which much  like in typescript would mean that module arguments that do not specify type constraints would be unchecked